backend/src/order.ts

```diff
@@ -354,162 +336,252 @@ export function getAllForShow(show_id: number): Promise<IAdminOrderListItem[]> {
-            log.error('Got an error from Paytrail', {error: err, responseBody: body, statusCode: response.statusCode, statusMessage: response.statusMessage,
-              requestHeaders: response.request.headers, requestBody: response.request.body, requestUri: response.request.uri, order_id: order_id});
-            reject(err);
-          }
-          var url = body.url;
-          db.query('update nk2_orders set payment_url = :url where id = :order_id',
-            {order_id: order_id, url: url});
-
-          resolve({url: url});
-        });
-      });
-    })
-    .catch((err) => {
-      log.error('Failed to prepare payment', {error: err, order_id: order_id});
-      throw err;
-    });
+  log.log('debug', 'Reject if status is not seats-reserved', {order_id});
+  if (res[0]['status'] !== 'seats-reserved'){
+    err = {name: "Validation error", message: "Invalid order status"}
+    log.error('Failed to prepare payment', {error: err, order_id:order_id})
+    await db.rollback(conn);
+    return reject(err);
+  }
+
+  log.log('debug', 'Set order status to payment-pending', {order_id});
+  [res, err] = await to(db.query('update nk2_orders set status = "payment-pending" where id = :order_id', {order_id}, conn));
+  if (err) {
+    log.error('Failed to prepare payment', {error: err, order_id: order_id});
+    await db.rollback(conn);
+    return reject(err);
+  }
+
+  log.log('debug', 'Commit payment status', {order_id});
+  [res, err] = await to(db.commit(conn));
+  if(err) {
+    log.error('Failed to prepare payment', {error: err, order_id: order_id});
+    return reject(err);
+  }
+
+  [order, err] = await to<IOrder>(get(order_id));
+  if (err) {
+    log.error('Failed to prepare payment', {error: err, order_id: order_id});
+    return reject(err);
+  }
+
+  const args: ICreateArgs = {
+    successRedirect: config.public_url + 'api/orders/' + order_id + '/success',
+    errorRedirect: config.public_url + 'api/orders/' + order_id + '/failure',
+    successCallback: config.public_url + 'api/orders/' + order_id + '/notify/success',
+    errorCallback: config.public_url + 'api/orders/' + order_id + '/notify/failure',
+  };
+
+  log.log('debug', 'Creating payment with handler', {order_id});
+
+  [handler, err] = await to(getPaymentHandler(order_id));
+  if (err) {
+    return reject(err);
+  }
+
+  [createResp, err] = await to<ICreateResponse>(handler.create(order, args));
+  if (err) {
+    log.error('Failed to prepare payment', {error: err, order_id: order_id});
+    return reject(err);
+  }
+
+  log.log('debug','Update payment status', {order_id})
+  await db.query('update nk2_orders set payment_id = :payment_id, payment_url = :payment_url where id = :order_id', {payment_id: createResp.payment_id, payment_url: createResp.payment_url, order_id});
+  //Hack here a little bit, we are only interested in error, not result value
+  if (err) {
```

Comment:
I think this "err" is the one you were using above (`[createResp, err] = await to<ICreateResponse>(handler.create(order, args));`)?